### PR TITLE
Observability

### PR DESCRIPTION
**Pull request by redhatHameed**
## Description

This is initial PR which will add metrics for the RHODA operator and enabled the configuration of scraping the metrics for local prometheus, then push for remote write to observatorium as show in the below flow diagram.
Screen Shot 2022-07-20 at 2 08 06 PM

##Task Done:

1. Install the Observability operator by using reconciler code as done for other operators.
2. Once the Observability operator installed creating the observability operator [CR](https://github.com/RHEcosystemAppEng/dbaas-operator/blob/observability/controllers/reconcilers/providers_installation/observability-operator-cr.yaml) with this it created observability related components with in dbaas namespace. 
3. Enabled the Configuration for scraping the metrics from RHODA Operator, with this it will create ServiceMonirtor prometheus CR with label 'app: dbaas-prometheus'.
4. Added metrics for inventory status and request time duration from the provider operator response 
5. Added the connection metrics status and request time duration to the from the provider operator response 
6. Added the instance metrics status and request time duration to the from the provider operator response 
7. Added metrics for platform installation and total duration taken to complete the installation.
8. Review the ADR for more details here [1](https://docs.google.com/document/d/1ol1l8Ad77Ab2i2YOICgsle2xBff6U16Eu562CBb-W-k/edit) and [2 ](https://docs.google.com/document/d/1Tv9AV4smstEI139O_k2tXPUFG6fZb01xs9Nju2jIg4w/edit#)

## Verification Steps

1. Checkout the PR locally, and build the image and deploy the RHODA operator.
2. Verify the RHODA operator install the Observability operator along with other provider operators.
3. Verify the CR for Observability operator created successfully and added extra pods and services with in dbaas operator namesapce.
4. created the different objects for inventory, connection, instance, 
5. Query the prometheus to see the metrics related with dbaas operator exists locally.
6. Currently remote write configuration is hardcoded add the the details into observability operator CR and verify the metrics are pushing into remote observatorium.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist (Definition of Done)

- [ ] All acceptance criteria specified in JIRA or in issue number have been completed
- [ ] Unit tests added that prove the fix is effective or the feature works 
- [ ] Documentation added for the feature
- [ ] all relevant tests are passing
- [ ] Code Review completed
- [ ] Verified independently by reviewer

### CODE REVIEW

File: docs/install-mso.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# Install and Deploy MSO instance
+___
+
+We have to deploy the MSO after installing the dbaas-operator on the cluster so that we can see and query the metrics from Prometheus UI.
```
[the remark this answers is not in the document]
Comment (redhatHameed) [reply docs/install-mso.md:4]: also the MSO operator renamed as observability operator so consider both same thing

[the remark this answers is not in the document]
Comment (jianrongzhang89) [reply docs/install-mso.md:4]: Sounds good. Please clean up the code then :)

**Comment by jianrongzhang89** [comment]
>> "5. Query the prometheus to see the metrics related with dbaas operator exists locally."

@redhatHameed  I did not see an OpenShift route created for prometheus. What's the URL for the prometheus? Thanks.

**Comment by redhatHameed** [comment]
> > > "5. Query the prometheus to see the metrics related with dbaas operator exists locally."
> 
> @redhatHameed I did not see an OpenShift route created for prometheus. What's the URL for the prometheus? Thanks.

you can create the route by going route menu and select the service `prometheus-operated` or 

use the oc port forward for the pod like `oc port-forward  prometheus-dbaas-operator-mso-0 28015:9090` and access locally  ->http://localhost:28015/metrics

**Comment by jianrongzhang89** [comment]
> > > > "5. Query the prometheus to see the metrics related with dbaas operator exists locally."
> > 
> > 
> > @redhatHameed I did not see an OpenShift route created for prometheus. What's the URL for the prometheus? Thanks.
> 
> you can create the route by going route menu and select the service `prometheus-operated` or
> 
> use the oc port forward for the pod like `oc port-forward prometheus-dbaas-operator-mso-0 28015:9090` and access locally ->http://localhost:28015/metrics

Sounds good.

**Comment by jianrongzhang89** [comment]
@redhatHameed  dbaas_platform_installation_status has version in the labels for all operators, but not for the dynamic plugins and telemetry plugin. Could the version information be added for them too? We can use the image version for them I think.

File: controllers/metrics.go [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+func PlatformInstallStart() Execution {
+	return Execution{
+		begin: time.Now(),
+	}
+}
+
+// PlatformStackInstallationMetric is used to log duration and success/failure
+func (e *Execution) PlatformStackInstallationMetric(version string) {
+	duration := time.Since(e.begin)
+	DBaasOperatorVersionInfo.With(prometheus.Labels{metricLabelVersion: version}).Set(0)
```
Comment (jianrongzhang89) [inline controllers/metrics.go:119]: Set the value to 0 so that it is consistent with other metrics (such as dbaas_platform_installation_status)?

**Comment by redhatHameed** [comment]
> @redhatHameed dbaas_platform_installation_status has version in the labels for all operators, but not for the dynamic plugins and telemetry plugin. Could the version information be added for them too? We can use the image version for them I think.

The easies way will be add another filed for version here and pass the metrics->https://github.com/RHEcosystemAppEng/dbaas-operator/blob/bb7952e7b2992cab31fcab38651ecc21af2eb9a7/controllers/reconcilers/constants.go#L44

**Comment by jianrongzhang89** [comment]
> > @redhatHameed dbaas_platform_installation_status has version in the labels for all operators, but not for the dynamic plugins and telemetry plugin. Could the version information be added for them too? We can use the image version for them I think.
> 
> The easies way will be add another filed for version here and pass the metrics->
> 
> https://github.com/RHEcosystemAppEng/dbaas-operator/blob/bb7952e7b2992cab31fcab38651ecc21af2eb9a7/controllers/reconcilers/constants.go#L44

Sounds a good idea :)

File: controllers/metrics.go (hunk 2 above)
Comment (redhatHameed) [reply controllers/metrics.go:119]: here already value setting 0- that does not meaning anything I think by default also value will be zero. Could you explain more about this, thanks

**Comment by jianrongzhang89** [comment]
@redhatHameed I created an inventory, instance and connection in my cluster. However the corresponding metrics do not show up. Can you take a look?  URL to my stack: https://console-openshift-console.apps.ty11em8p.eastus.aroapp.io/

Screen Shot 2022-07-21 at 10 56 41 AM

**Comment by redhatHameed** [comment]
> @redhatHameed I created an inventory, instance and connection in my cluster. However the corresponding metrics do not show up. Can you take a look? URL to my stack: https://console-openshift-console.apps.ty11em8p.eastus.aroapp.io/
> 
> Screen Shot 2022-07-21 at 10 56 41 AM

refresh ?

File: controllers/metrics.go (hunk 2 above)
Comment (jianrongzhang89) [reply controllers/metrics.go:119]: Sorry, I meant 1 instead as dbaas_platform_installation_status is set to 1 for the existence of such metrics.

**Comment by jianrongzhang89** [comment]
> > @redhatHameed I created an inventory, instance and connection in my cluster. However the corresponding metrics do not show up. Can you take a look? URL to my stack: https://console-openshift-console.apps.ty11em8p.eastus.aroapp.io/
> > Screen Shot 2022-07-21 at 10 56 41 AM
> 
> refresh ?

Refreshed the screen and now the metrics show up.

**Comment by jianrongzhang89** [comment]
I see there are two metrics for dbaas_instance_phase and the only difference is that one has provider account in it. Why do we need two metrics here?
Screen Shot 2022-07-21 at 11 32 10 AM

**Comment by redhatHameed** [comment]
> I see there are two metrics for dbaas_instance_phase and the only difference is that one has provider account in it. Why do we need two metrics here? Screen Shot 2022-07-21 at 11 32 10 AM

should be one,  looking into

File: controllers/metrics.go (hunk 2 above)
Comment (redhatHameed) [reply controllers/metrics.go]: done

**Comment by redhatHameed** [comment]
> > I see there are two metrics for dbaas_instance_phase and the only difference is that one has provider account in it. Why do we need two metrics here? Screen Shot 2022-07-21 at 11 32 10 AM
> 
> should be one, looking into

Fixed now

**Review by jeremyary — APPROVED** [review]

**Pull request merged**